numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -2344,10 +2344,12 @@ PyArray_ConvertToCommonType(PyObject *op, int *retn)
         }
 
         mps[i] = (PyArrayObject *)PyArray_FROM_O(tmp);
-        Py_DECREF(tmp);
         if (mps[i] == NULL) {
+            Py_DECREF(tmp);
             goto fail;
         }
+        npy_mark_tmp_array_if_pyscalar(tmp, mps[i], NULL);
```

Comment:
Bit off-topic for here, but should this generally be done as part of `PyArray_FROM_O`? Presumably, that has all the necessary information to make it very quick.